### PR TITLE
updates for using ex_aws as s3 store

### PR DESCRIPTION
**Pull request by benwilson512**

### CODE REVIEW

File: lib/hex_web/store/s3.ex [hunk 1]
```diff
@@ -90,33 +87,15 @@ defmodule HexWeb.Store.S3 do
...
-    new_keys = result[:contents]
-    all_keys = new_keys ++ keys
-
-    if result[:is_truncated] do
-      list_all(bucket, prefix, List.last(new_keys), all_keys)
-    else
-      all_keys
-    end
+    Application.get_env(:hex_web, bucket)
+    |> S3.stream_objects(prefix: prefix)
```
Comment (benwilson512) [inline lib/hex_web/store/s3.ex]: This is much simpler now. Will need to see if the maps that this streams are compatible.

File: lib/hex_web/store/s3.ex [hunk 2]
```diff
@@ -90,33 +87,15 @@ defmodule HexWeb.Store.S3 do
     HexWeb.Plug.redirect(conn, url)
   end
 
-  def upload(bucket, path, headers, data) do
+  def upload(bucket, path, data, opts \\ []) do
```
Comment (benwilson512) [inline lib/hex_web/store/s3.ex:89]: Putting opts last and making them optional really simplifies a lot of the calls to it in this file.

File: lib/hex_web/store/s3.ex [hunk 3]
```diff
@@ -90,33 +87,16 @@ defmodule HexWeb.Store.S3 do
...
-    all_keys = new_keys ++ keys
-
-    if result[:is_truncated] do
-      list_all(bucket, prefix, List.last(new_keys), all_keys)
-    else
-      all_keys
-    end
+    Application.get_env(:hex_web, bucket)
+    |> S3.stream_objects(prefix: prefix)
+    |> Stream.map(&Map.get(&1, :key))
```
Comment (benwilson512) [inline lib/hex_web/store/s3.ex:99]: This is much simpler now.

**Comment by benwilson512** [comment]
How do you want to handle errors? ExAws returns `{:ok, results}` or `{:error, reason}` tuples, so at a minimum stuff like https://github.com/benwilson512/hex_web/blob/switch-to-ex_aws/lib/hex_web/store/s3.ex#L11 needs to look more like

```
    {:ok, result} = S3.get_object(bucket, key)
    result
```

It looks like based on the existing code that it will retry `@max_attempts` times at which point you'll get a match error. https://github.com/hexpm/hex_web/blob/master/lib/hex_web/store/s3.ex#L143

If that's what you want, then you could basically just do `{:ok, _} = S3.whatever`. Alternatively I've been meaning to create bang functions for everything that basically shortcut that process.

Thoughts?

**Comment by benwilson512** [comment]
@ericmj may have opinions here as well.

**Comment by ericmj** [comment]
A match error is fine, I get mails for each exception so I can manually review the error.

**Comment by ericmj** [comment]
Everything looks great to me, we just needs to make sure we raise on errors.

File: lib/hex_web/store/s3.ex [hunk 4]
```diff
@@ -55,17 +55,13 @@ defmodule HexWeb.Store.S3 do
...
-      "." <> ext ->
-        mime = Plug.MIME.type(ext)
-        headers = ["content-type": mime]
-      "" ->
-        headers = []
+    opts = case Path.extname(path) do
+      "." <> ext -> [content_type: Plug.MIME.type(ext)]
+      ""         -> []
     end
+    |> Keyword.put(:cache_control, "public, max-age=1800")
```
Comment (benwilson512) [inline lib/hex_web/store/s3.ex:62]: Note that headers are normal looking atoms.

File: lib/hex_web/store/s3.ex [hunk 5]
```diff
@@ -90,32 +86,16 @@ defmodule HexWeb.Store.S3 do
...
-
-    if result.is_truncated == "true" do
-      list_all(bucket, prefix, List.last(new_keys), all_keys)
-    else
-      all_keys
-    end
+    Application.get_env(:hex_web, bucket)
+    |> S3.stream_objects!(prefix: prefix)
+    |> Stream.map(&Map.get(&1, :key))
   end
```
Comment (benwilson512) [inline lib/hex_web/store/s3.ex:100]: A much simpler list_all implementation ;)

**Comment by mschae** [comment]
:+1: very cool

**Pull request merged**